src/Solver/Pipeline/GenericPipeline.h

```diff
@@ -0,0 +1,233 @@
+        this->end += stepSize;
+        ++currentIteration;
+        clamp();
+        return *this;
+      }
+
+    private:
+      void clamp() {
+        begin = (begin > limit) ? limit : begin;
+        end = (end > limit) ? limit : end;
+      }
+    };
+
+    void init(size_t size) {
+      const auto currBatchSize = tuner.getBatchSize();
+      for (auto& range: ranges) {
+        range = Range(currBatchSize, size);
+      }
+      numTotalIterations = (size + currBatchSize - 1) / currBatchSize;
+      numFullPipeIterations = (numTotalIterations > TailSize) ? numTotalIterations - TailSize : 0;
+      stageTiming = decltype(stageTiming){};
+    }
+
+    void fill() {
+      for (unsigned i = 0; i < TailSize; ++i) {
+        for (unsigned stage = 0; stage < (i + 1); ++stage) {
+          if (ranges[stage].currentIteration < numTotalIterations) {
+            execute(stage, ranges[stage]);
+            ranges[stage]++;
+            checkToFinalize(stage, ranges[stage].currentIteration);
+          }
+        }
+      }
+    }
+
+    void run() {
+      // reduce numFullPipeIterations by 1 to handle a corner case in the `drain` stage
+      const size_t length = (numFullPipeIterations > 0) ? numFullPipeIterations - 1 : 0;
+      for (size_t i = 0; i < length; ++i) {
+        for (unsigned stage = 0; stage < NumStages; ++stage) {
+          execute(stage, ranges[stage]);
+          ranges[stage]++;
+        }
+      }
+    }
+
+    void drain() {
+      for (unsigned i = 0; i < TailSize + 1; ++i) {
+        for (unsigned stage = 0; stage < NumStages; ++stage) {
+          if (ranges[stage].currentIteration < numTotalIterations) {
+            execute(stage, ranges[stage]);
+            ranges[stage]++;
+            checkToFinalize(stage, ranges[stage].currentIteration);
+          }
+        }
+      }
+    }
+
+    void clean() {
+      if (resetAfterRun) {
+        for (auto& callBack: callBacks) {
+          callBack = nullptr;
+        }
+      }
+    }
+
+    void execute(unsigned stage, Range range) {
+      assert(callBacks[stage] != nullptr && "call back has not been setup");
+      stopwatch.start();
+      (*callBacks[stage])(range.begin, range.size(), range.currentIteration);
+      stageTiming[stage] += stopwatch.stop();
+    }
+
+    void checkToFinalize(unsigned stage, size_t currentIteration) {
+      assert(callBacks[stage] != nullptr && "call back has not been setup");
+      if (currentIteration == numTotalIterations) {
+        stopwatch.start();
+        callBacks[stage]->finalize();
+        stageTiming[stage] += stopwatch.stop();
+      }
+    }
+
+    size_t numTotalIterations{0};
+    size_t numFullPipeIterations{0};
+    std::array<Range, NumStages> ranges;
+    std::array<double, NumStages> stageTiming{};
+    std::array<PipelineCallBack*, NumStages> callBacks{nullptr};
+    bool resetAfterRun{true};
+    TunerT tuner;
+    Stopwatch stopwatch{};
+
+    friend class seissol::unit_test::PipelineTest;
```

Comment:
Ok. There are two opinions about friend classes for unit testing. The ones who say "don't do it" mean that a class implementation may change in the future which will definitely lead to some changes in tests. I agree that friends create a tight coupling between classes and their tests. The ones who say "it is ok" mean that they want to avoid writing lot's of getter methods that nobody will use except test classes. In my case, a client only need two public methods, namely: `run` and `registerCallBack`.

I am personally in between of these two opinions because I agree with both of them. There is no strict preference of one approach over another one in the discussion that you attached. People say that sometimes it is ok to have friends for testing. 

In short, I need to access 2 private attributes of a pipeline: `numFullPipeIterations` and `numTotalIterations`. I can write getters for them.

What do you think?

Reply:
Imo unit tests should only test the public interface

Reply:
Done, Friend classes have already been removed